src/ripple/app/ledger/Ledger.cpp

```diff
@@ -1000,6 +972,63 @@ Ledger::unchecked_replace(
     auto const ours = std::move(sle);
 }
 
+std::size_t
+Ledger::txCount() const
+{
+    // Always zero for closed ledgers.
+    return 0;
+}
+
+bool
+Ledger::txExists (uint256 const& key) const
+{
+    return txMap().hasItem (key);
+}
+
+void
+Ledger::txInsert (uint256 const& key,
+    std::shared_ptr<Serializer const
+        > const& txn, std::shared_ptr<
+            Serializer const> const& metaData)
+{
+    if (metaData)
+    {
+        // low-level - just add to table
+        Serializer s(txn->getDataLength () +
+            metaData->getDataLength () + 16);
+        s.addVL (txn->peekData ());
+        s.addVL (metaData->peekData ());
+        auto item = std::make_shared<
+            SHAMapItem const> (key, std::move(s));
+        if (! txMap().addGiveItem
+                (std::move(item), true, true))
+            LogicError("duplicate_tx: " + to_string(key));
```

Comment:
Just a note that `LogicError` does not write the message to the log. I'm guessing this is because it's going to core dump and the message is unlikely to be written to the log? If that isn't the case (I'm not familiar with how the log is implemented) then we should write the message.